nvflare/app_common/tie/process_mgr.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import shlex
+import subprocess
+import sys
+import threading
+
+
+class ProcessManager:
+    def __init__(self):
+        """Constructor of ProcessManager.
+        ProcessManager provides methods for managing the lifecycle of a subprocess (start, stop, poll), as well
+        as the handling of log file to be used by the subprocess.
+
+        NOTE: the methods of ProcessManager are not thread safe.
+
+        """
+        self.process = None
+        self.log_file = None
+        self.log_prefix = None
+
+    def start(
+        self,
+        command: str,
+        log_prefix: str = None,
+        cwd=None,
+        env=None,
+        log_file=None,
+        log_stdout=True,
+    ):
+        """Start the subprocess.
+
+        Args:
+            command: the command to be executed
+            log_prefix: prefix to be prepended to log message when writing to stdout
+            cwd: current work dir for the process to be started
+            env: system env for the process
+            log_file: log file for the process. It can be: a file object, a path to the file, or None.
+                If None, no log file will be used.
+            log_stdout: whether to output log messages to stdout.
+
+        Returns: None
+
+        """
+        self.log_prefix = log_prefix
+
+        lf = log_file
+        if log_file and isinstance(log_file, str):
+            lf = open(log_file, "a")
+            self.log_file = lf
+
+        if lf and log_stdout:
+            stdout = subprocess.PIPE
+        elif lf and not log_stdout:
+            stdout = lf
+        else:
+            stdout = None
+
```

Comment:
If users don't specify log_file, and set log_stdout = True.
We will set stdout=None, and nothing gets printed out?

Reply:
Actually I'm not sure. None is the default value of stdout for popen and it seems to print it to sys.stdout by default.
Maybe we should manage it through our code. I tried to avoid the use of another thread.
We can wait a bit before doing anything because in reality both log file and log_stdout will be used.

Reply:
BTW, I changed to manage all of these thru our code so will leave no ambiguity.